### PR TITLE
Realization Configuration Generator

### PR DESCRIPTION
**Pull request by jdmattern-noaa**
Generates Realization Configuration JSON file for a given set of catchments. It randomly selects between two formulations for each catchment, assigns corresponding params, and outputs a full realization JSON file.

## Notes

-

## Todos

-

## Checklist

- [ ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

## Testing checklist (automated report can be put here)

1.

### Target Environment support

- [ ] Linux

### CODE REVIEW

File: data/config_file_generators/realization_config_generator.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+formulations for each catchment given a catchment data geojson file as input. 
+
+Script and command line arg format:
+realization_config_generator.py -i <path and file name to catchment data geojson file>
+-o <file name of config output file>
+
+Example:
+python realization_config_generator.py -i ../catchment_data.geojson -o realization_config.json
+
+TODO: Determine method to pass each catchment's forcing file to catchment JSON object
```
Comment (jdmattern-noaa) [inline data/config_file_generators/realization_config_generator.py:13]: @donaldwj @hellkite500 I am unclear about what the current consensus is for feeding the forcing for each catchment for the demo. Is it through ngen or BMI? If it is through ngen, then I can find a method to pass each catchment's forcing file to this config.

File: data/config_file_generators/realization_config_generator.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+"""
+
+import sys, getopt
+import geopandas as gpd
+import random
+import json
+
+#Set the names of the formulations
+global_formulation = "tshirt_c"
+formulation_1 = "top_model"
```
[the remark this answers is not in the document]
Comment (jdmattern-noaa) [reply data/config_file_generators/realization_config_generator.py:23]: Thanks!

File: data/config_file_generators/realization_config_generator.py (hunk 1 above)
Comment (jdmattern-noaa) [reply data/config_file_generators/realization_config_generator.py:13]: The discussion this morning clarified things for me more on this. The existing structure should allow to set forcing at the top, and each catchment would use it.

**Comment by robertbartel** [comment]
@jdmattern-noaa, this is dependent on the outcome from #251, but you are likely going to need to add `registration_function` and `variable_names_map` parameters for each BMI C `formulation` entry.   I have an example with CFE below, but a TOPMODEL will need to be altered similarly, though I'm not exactly sure of all the values yet.

```
    "cat-27": {
            "formulations": [
                {
                    "name": "bmi_c",
                    "params": {
                        "model_type_name": "bmi_c_cfe",
                        "library_file": "./extern/alt-modular/cmake_am_libs/libcfebmi.so",
                        "forcing_file": "./data/forcing/cat-27_2015-12-01 00_00_00_2015-12-30 23_00_00.csv",
                        "init_config": "./data/bmi/c/cfe/cat_27_bmi_config.txt",
                        "main_output_variable": "Q_OUT",
                        "registration_function": "register_bmi_cfe",
                        "variable_names_map" : {
                            "water_potential_evaporation_flux" : "potential_evapotranspiration",
                            "atmosphere_water__liquid_equivalent_precipitation_rate" : "precip_rate",
                            "atmosphere_air_water~vapor__relative_saturation" : "SPFH_2maboveground",
                            "land_surface_air__temperature" : "TMP_2maboveground",
                            "land_surface_wind__x_component_of_velocity" : "UGRD_10maboveground",
                            "land_surface_wind__y_component_of_velocity" : "VGRD_10maboveground",
                            "land_surface_radiation~incoming~longwave__energy_flux" : "DLWRF_surface",
                            "land_surface_radiation~incoming~shortwave__energy_flux" : "DSWRF_surface",
                            "land_surface_air__pressure" : "PRES_surface"
                        },
                        "uses_forcing_file": true
                    }
                }
            ],
            "forcing": {
                "path": "./data/forcing/cat-27_2015-12-01 00_00_00_2015-12-30 23_00_00.csv"
            }
        },
```

**Comment by robertbartel** [comment]
Regarding the necessary changes to the realization config format for TOPMODEL formulation, it looks like `variable_names_map` should be just these two:

```
                       ...
                        "variable_names_map" : {
                            "water_potential_evaporation_flux" : "potential_evapotranspiration",
                            "atmosphere_water__liquid_equivalent_precipitation_rate" : "precip_rate"
                        },
                        ...
```

Also, `registration_function` should be changed to `register_bmi_topmodel`.

**Comment by jdmattern-noaa** [comment]
@robertbartel Do you know what the library file for top model would be?
                        "library_file":

**Comment by jdmattern-noaa** [comment]
@robertbartel Do you know if the "init_config": is also passed through BMI to top model? I know they are consolidating their inputs into one dat file.

**Comment by hellkite500** [comment]
@jdmattern-noaa did you update this with any required changes discovered from the demo?

**Comment by jdmattern-noaa** [comment]
> @jdmattern-noaa did you update this with any required changes discovered from the demo?



> @jdmattern-noaa did you update this with any required changes discovered from the demo?

@hellkite500 Yes, I made the updates that we discussed to where there doesn't need to be a string replacement for the model name and cleaned up other items. Is a utilities directory at the top level a good place for this script? Also, do you have that script you mentioned in the meeting the other day to run on this?

**Review by robertbartel — APPROVED** [review]

**Pull request merged**